EIPS/eip-6963.md

```diff
@@ -0,0 +1,234 @@
+
+In this proposal, we provide a solution that optimitizes for interoperability of multiple wallet providers and enables fairer competition by lowering the barrier to entry for new wallet providers to improve the user experience on Ethereum.
+
+This is achieved by introducing a set of window events to provide a two-way communication protocol between Ethereum libraries and injected scripts provided by browser extensions thus enabling users to select their wallet of choice.
+
+## Specification
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+### Provider Info
+
+Each wallet provider will be exposed with the following interface `EIP6963ProviderInfo` that will be used to display to the user:
+
+```typescript
+/**
+ * Represents the assets needed to display a wallet
+ */
+interface EIP6963ProviderInfo {
+  walletId: string;
+  uuid: string;
+  name: string;
+  icon: string;
+}
+```
+
+The values in the `EIP6963ProviderInfo` MUST be used as follows:
+
+- walletId - globally unique identifier of the wallet provider (eg. `io.dopewallet.extension` or `awesomewallet`)
+- uuid - locally unique of the wallet provider (UUIDv4 compliant)
+- name - human-readable name of the wallet provider (e.g. `DopeWalletExtension` or `AwesomeWallet`)
+- icon - uri encoded image (RFC-3986 compliant, image must be squared with 96x96px minimum resolution)
+
+### Announced Provider
+
+The wallet provider will also expose their own [EIP-1193](./eip-1193.md) provider interface in parallel with the provider info provided above in the following interface `EIP6963AnnouncedProvider`
+
+```typescript
+interface EIP6963AnnouncedProvider {
+  info: EIP6963ProviderInfo;
+  provider: EIP1193Provider;
+}
+```
+
+The `EIP1193Provider` interface is documented at [EIP-1193](./eip-1193.md) and can be used to override the `window.ethereum` object once the user as explicitly selected it.
+
+### Window Event
+
+Different wallet providers might inject their scripts at different timeframes plus there is no guarantees that the Ethereum library in the web page will be loaded before all injected scripts are present in the page.
+
+To emit events it must use `window.postMessage` and to observe events it must use `window.addEventListener`. There will be two event interfaces:
+
+```typescript
+// Requesting an EVM provider
+interface EIP6963RequestProviderEvent extends MessageEvent {
+  data: {
+    eventName: "eip6963:requestProvider";
+  };
+}
+
+// Annoucing an EVM provider
+interface EIP6963AnnounceProviderEvent extends MessageEvent {
+  data: {
+    eventName: "eip6963:announceProvider";
+    announcedProvider: EIP6963AnnouncedProvider;
```

Comment:
Unfortunately I don’t think `postMessage()` can be used to pass an EIP-1193 object as it contains functions, [which will throw a serialization error](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm#things_that_dont_work_with_structured_clone) if we attempt to pass it to `postMessage()`. /cc @MicahZoltu

Reply:
It depends on the browser, but generally you should stick to anything that can JSON serialize (no functions, no bigints, no Uint8Array) as that is the lowest common denominator (Chromium).

Reply:
`BigInt` and `Uint8Array` (or any typed array with certain limitations) must be supported by browsers (see the [relevant part](https://html.spec.whatwg.org/multipage/structured-data.html#structuredserializeinternal) of the HTML spec) to serialize objects passed to `postMessage()`. But yes the issue is that functions are not allowed, making EIP-1193 providers unserializable with this algorithm, and incompatible with `postMessage()`.

Reply:
Hmmm, I must be thinking of some other communication channel then, sorry!

Reply:
Personally I also find dispatchEvent to be a cleaner approach but it does reduce the interoperability for this standard

Using postMessage not only makes it compatible with browser extensions but also iframes

Reply:
It would be much easier if EIP-1193 was using a message-based API rather than shipping an object with its own methods… maybe one day :sweat_smile:

As things are, I only see one way to keep using `postMessage`, which would be to communicate the name of the `ethereum` object attached to `window` rather than the object itself… or we could reuse the uuid for this, e.g.:

```ts
function onPageLoad() {
  let provider: EIP1193Provider;
  const uuid = "350670db-19fa-4704-a166-e52e178b59d2";

  window.ethereum = provider;

  // we attach the provider using the uuid
  window[`ethereum_eip6963_${uuid}`] = provider;

  function announceProvider() {
    const info: EIP6963ProviderInfo = {
      walletId: "org.example.wallet",
      uuid,
      name: "Example Wallet",
      icon: "https://wallet.example.org/icon.png",
    };
    const announcedProvider = { info, provider };
    window.postMessage({
      eventName: "eip6963:announceProvider",

      // we only communicate the uuid here
      announcedProviderUuid: uuid,
    });
  }

  window.addEventListener("message", (event: MessageEvent) => {
    if (event.data.eventName === "eip6963:requestProvider") {
      announceProvider();
    }
  });

  announceProvider();
}
```

But it would require to have an access to `window` to inject the provider no matter what, which kind of defeats the purpose of using `postMessage`.

So yeah reverting back to `dispatchEvent()` is probably the best option we have, since implementing the EIP-1193 methods require to rely on the current context no matter what, in order to communicate with the wallet.

Reply:
Requiring the app to inject into the page's context defeats the purpose of using `window.postMessage`.  The primary argument in favor of using `window.postMessage` is that it is the only communication channel that works for both extensions and iframes and any other kind of sandboxing or containment.  As long as the app has to inject content into the page's context, then extensions will have to do the hacky thing where they inject a script into the DOM which then executes in the page's context to write into the page (something that should generally not be done).

Reply:
+1 for switching to a message-based system. If we're already going with `postMessage` for announcing the provider itself, might as well go all the way and replace all the EIP-1193 functions with standardized messages too.

Reply:
After doing some testing with both `postMessage` and `dispatchEvent`... I think the best solution is to revert to `dispatchEvent`

Using postMessage with the current spec will break as flagged by @bpierre (screenshot below)
<img width="756" alt="Screenshot 2023-05-14 at 10 40 38" src="https://github.com/ethereum/EIPs/assets/10136079/e3cd4124-6d58-450c-833c-0783b85a9d8c">

The alternatives are the following:
1. use postMessage with a window map using uuid
2. use postMessage exclusively without EIP-1193 interface
3. revert to dispatchEvent instead of postMessage

Option 1 with the window map using uuid is a simple solution to this problem but defies the whole purpose of postMessage since it will require wallet providers to have window access. If they have window access then they can also use dispatchEvent.

Option 2 would require us to remove the dependency on EIP-1193 spec and instead make all providers use window.postMessage exclusively. IMHO this adds too much friction from both Ethereum libraries and Wallet providers to adapt

Option 3 would simply replace postMessage with dispatchEvent which would immediately make this EIP feasible to be implemented with little to no breaking changes in both Ethereum libraries and Wallet providers (screenshot below)

<img width="758" alt="Screenshot 2023-05-14 at 10 50 18" src="https://github.com/ethereum/EIPs/assets/10136079/e8b33538-a998-4247-8088-239ab659bcc5">

The impact that Option 3 has completely outweights the very low cost it has in terms of developer time in the community to adopt across multiple Ethereum libraries and Wallet providers